flips/20211129-cadence-mutability-restrictions.md

```diff
@@ -0,0 +1,417 @@
+contracts reviewed in order to deploy them to Flow. By removing one possible way that 
+users could write unsafe smart contracts, we make it easier for everyone to trust the
+correctness of contracts deployed to Flow. 
+
+## Design Proposal
+
+### Explanation 
+
+Cadence controls where and to what extent variables can be read and written using a combination of
+access modifiers and declaration kinds, as described [here](https://docs.onflow.org/cadence/language/access-control/).
+Of note is that the `let` kind does not allow fields to be written to in any scope, whereas `var` allows them
+to be written in the "Current and Inner" scopes; that is, the scope in which the field was declared, and any scopes
+contained within that scope. 
+
+However, simply writing to a field directly is not the only way in which one can modify a value. Consider the following example:
+
+```
+pub struct Foo {
+    pub let x : [Int]
+
+    init() {
+        self.x = [3];
+    }
+}
+
+pub fun bar() {
+    let foo = Foo()
+    foo.x = [0] // writes to x, not allowed
+    foo.x[0] = 0; // does not write to x, also not allowed
+}
+```
+
+Cadence also restricts the scopes in which an array or dictionary field can be modified (or "mutated"). Examples of 
+mutating operations include an indexed assignment, as in the above example, as well as calls to the `append` or `remove`,
+methods of arrays, or the `insert` or `remove` methods of dictionaries. These operations can only be performed on a field
+in the current and inner scopes, the same contexts in which the field could be written to if it were a `var`. So the following 
+would typecheck:
+
+```
+pub struct Foo {
+    pub let x : [Int]
+
+    init() {
+        self.x = [3];
+    }
+
+    pub fun addToX(i: Int) {
+        self.x.append(i)
+    }
+}
+```
+
+while the following would not:
+
+```
+pub struct Foo {
+    pub let y : [Int]
+
+    init() {
+        self.y = [3];
+    }
+}
+
+pub fun addToY(foo: Foo, i: Int) {
+    foo.y.append(i)
+}
+```
+
+This prevents external code from mutating the values of fields it can read from your contract. Consumers
+of your contract may read the values in a `pub let` or `pub var` field, but cannot change them in any way. 
+
+If you wish to allow other code to update or modify a field in your contract, you may expose a method 
+to do so, like in the example above with `addToX`, or you may use the `pub(set)` access mode, which 
+allows any code to mutate or write to the field it applies to. 
+
+### Implementation Details 
+
+This change adds a new error, the `ExternalMutationError`, which is raised when a field 
+is mutated outside of the context in which it was defined. The error message will also
+suggest that the user instead use a setter or modifier method for that field.
+
+Specifically, the error is emitted whenever a user attempts to perform an 
+index assignment on a member that is not either declared with the `pub(set)` 
```

Comment:
Does the proposal specifically only handle arrays and dictionaries, and the indexing assignment, or does it also address field assignments on composite values?
For example:

```rust
pub struct Foo {
  pub var name: String
  init() {
    self.name = "initial"
  }
}

pub struct Bar {
  pub let foo: Foo
  init() {
    self.foo = Foo()
  }
}

let bar = Bar()
bar.foo.name = "new"  // is this also rejected?
```

Reply:
I think this is the case : 
`pub struct Foo {
  pub(set) var name: String
  init() {
    self.name = "initial"
  }
}

pub struct Bar {
  pub let foo: Foo
  init() {
    self.foo = Foo()
  }
}

let bar = Bar()
bar.foo.name = "new"  // is this also rejected?`

Reply:
@dsainati1 Sorry, I wrote the code in the comment field and missed the essential part, like @bluesign corrected, `Foo.name` should be `pub(set)`. 

The point I'm trying to make here: Just like the proposal is currently proposing to improve working with containers such as arrays and dictionaries, which are mutable types, composites can also be mutable.

Reply:
Yea this is an interesting point, and one where I think there is room for more discussion on the design. It is certainly possible that we should restrict writes to composites like this: i.e. if any of the members in an `x.y.z = e` chain are declared with something either than `pub(set) var`, we might not allow them to be mutated. However, given that `pub(set) var`s must be explicitly declared by users, I don't believe they as likely to be accidentally mutated. 

In my opinion, I think we can use the same behavior for mutations we currently use for writes, wherein a `pub(set)` field can still be written to even if it is a member of a composite that is stored in a `let` field. However, I am curious what arguments there are for a different approach.

Reply:
Would this apply to calling functions on composite fields also? I don't think it would, but maybe something worth considering

Reply:
I like the proposal, but it seems like it only addresses part of the general problem, i.e. only focuses on arrays and dictionaries, but composite values are also containers. 

Then again, maybe it is good to keep the scope of this proposal narrow, and we can extend the new behaviour to composites in a follow-up proposal.

Reply:
I'm really on the fence here, and the example in this document really makes it a tough call.

```
    foo.y.append(1) // invalid, field `y` is let and cannot be mutated
```

In this case, I'm not assigning to `y`, I'm calling a method on `y`. If `y` were a user-defined type with a method called `append()`, this would NOT be an error. To make it an error just because `y` is an language-defined collection type is pretty inconsistent.

OTOH- Containers are the place where this kind of "gotcha" is most likely to be a problem, so maybe it's worth introducing an inconsistency temporarily until we can find a better solution?

I don't feel confident enough to push this discussion in either direction, to be honest.

Reply:
Looking at this and @turbolent's comments below, the consensus seems to be that this FLIP does not address 100% of the footgun problem, but does hit the most common cases. If we want to handle the parts of this "gotcha" case, as I see it we have a couple options:

1) We can update the FLIP to handle the nested-composite mutation case. There are a couple ways I can see to to do this: we add a `mutating` keyword (or similar) to method definitions, and prevent such methods from being called on fields declared with a `let`. We would also have the option of enforcing that `mutating` methods are the only way to modify internal state (like Swift does), or we could choose to allow the modifier to be used at the developer's discretion, which is easier but also less accurate. 

2) We merge the FLIP now and attempt to address the nested-composite case later. The difficulty here is that any later solution to the nested-composite case would have to be a non-breaking change. Adding a `mutating` keyword as described above, but only restricting where `mutating` methods are used rather than also requiring certain methods to be marked as `mutating` is one way to address this without a breaking change, but it certainly isn't a perfect solution.

3) We merge the FLIP as-is, and just live with the inconsistency/incompleteness. 

My personal preference is for option 2, as long as we can come up with a non-breaking solution that we can apply later that is also a satisfying way to solve the problem. Interested to see if anybody has any thoughts on non-breaking fixes for this.

Reply:
I'm fine with just not allowing anything on the array and dictionary fields, since like Dete said, that is the place where the vast majority of people make their mistakes. The inconsistency doesn't bother me much.